### DISCUSSION TITLE
KafkaProducerMessageHandler to detect issues/errors coming from kakfa-client

### DISCUSSION THREAD

**Discussion by yevsh** [spring-projects/spring-integration#8637] [General]
Sometimes there are  "error" network/kafka cluster errors - that can be only detected when we see in logs something like:

`"INFO","logger":"org.apache.kafka.clients.NetworkClient","message":"[Producer clientId=producer-1] Node 2 disconnected`.

The problem KafkaTemplate/KafkaProducerMessageHandler can't identify it as error, and data can be lost.

Need some reliable solution for this.

**Comment by artembilan** [comment]
Since that is just an INFO from the Kafka Client, then there is indeed nothing we can do on our side: https://stackoverflow.com/questions/73823931/node-disconnected-errors-in-kafka-streams-api.

It is probably better to ask such a question on Apache Kafka community.

**Reply by garyrussell** [reply]
To avoid data loss, you can set `sync` and the thread will block until the send is successful or throws an exception, or you can configure a `sendFailureChannel` to asynchronously receive an error message for the failure.

https://github.com/spring-projects/spring-integration/blob/8b004e9ec2be349cd112c84ba7075e84f1eaa232/spring-integration-kafka/src/main/java/org/springframework/integration/kafka/outbound/KafkaProducerMessageHandler.java#L306-L315

https://github.com/spring-projects/spring-integration/blob/8b004e9ec2be349cd112c84ba7075e84f1eaa232/spring-integration-kafka/src/main/java/org/springframework/integration/kafka/outbound/KafkaProducerMessageHandler.java#L268-L276

**Comment by yevsh** [comment]
did as suggested :

```
 handler.setSync(true);
 handler.setSendFailureChannelName("failureChannel");
```

Then,  KafkaTemplate's doSend() , the exception throw was triggered: (2.8.11 - if it's important)

```
 catch (ExecutionException e) {
	 throw new KafkaException("Send failed", e.getCause()); // NOSONAR, stack trace
 }
```
with error:
`Caused by: org.springframework.kafka.KafkaException: Send failed; nested exception is org.apache.kafka.common.errors.TimeoutException: Topic xxxxxxx not present in metadata after 10000 ms.`

but message didn't move to failureChannel :

```
@ServiceActivator(inputChannel = "failureChannel")
    public void handleFailure(Message<?> message) {
}
```

**Reply by artembilan** [reply]
The logic there is like this:
```
	if (failureChannel != null || metadataChannel != null) {
			future.whenComplete((sendResult, exception) -> {
```

Which is applied to the `Future` returned by the `KafkaTemplate.send()`:
```
sendFuture = this.kafkaTemplate.send(producerRecord);
...
processSendResult(message, producerRecord, sendFuture, getSendSuccessChannel());
```

But turns out you gort a direct exception from that `.kafkaTemplate.send()` not a `Future`:
```
if (sendFuture.isDone()) {
			try {
				sendFuture.get();
			}
			catch (InterruptedException e) {
				Thread.currentThread().interrupt();
				throw new KafkaException("Interrupted", e);
			}
			catch (ExecutionException e) {
				throw new KafkaException("Send failed", e.getCause()); // NOSONAR, stack trace
			}
		}
```

Therefore that `failureChannel` is not called.
You need to look into a `RequestHandlerRetryAdvice` to catch that `KafkaException` and retry. If that is a logic you are looking for...

**Reply by garyrussell** [reply]
<strike>

Actually, with `sync`, the timeout is thrown by this code.

https://github.com/spring-projects/spring-integration/blob/c9023d114b653d185f8cce3cce2ff9b376249fe1/spring-integration-kafka/src/main/java/org/springframework/integration/kafka/outbound/KafkaProducerMessageHandler.java#L710-L731


However, you don't need to do both; sync for exceptions to be thrown to the caller async for receiving the error asynchronously.

Note that even with async, some exceptions can still be thrown to the caller by the code that @artembilan referenced (e.g. message too large exceptions).
</strike>

**Reply by garyrussell** [reply]
<strike>However, you should have received the error message in that case (see the code just above that).

Did you configure a consumer on the failure channel? </strike>

**Comment by garyrussell** [comment]
My apologies to @artembilan, in this case, the exception **is** thrown by the code he cited and will be thrown to the caller, regardless of the `sync` setting. There is no failure message sent for this exception.

@artembilan Do you think we should add a try/catch (and rethrow) block around this code:

https://github.com/spring-projects/spring-integration/blob/c9023d114b653d185f8cce3cce2ff9b376249fe1/spring-integration-kafka/src/main/java/org/springframework/integration/kafka/outbound/KafkaProducerMessageHandler.java#L517-L530

So we consistently send the error message?

On the other hand, if it is a direct channel, the caller will directly get the exception; if async, the exception will go to an error channel (if configured).

**Reply by artembilan** [reply]
I think, for consistency, we need to fix `KafkaTemplate` and don't throw an exception, but rather call `CompletableFuture.completeExceptionally(ex)`.
This way the consumer of that code would always have only one way to react to the output of `send()` method call - hook into returned `CompletableFuture`.

**Reply by garyrussell** [reply]
We can't change that behavior in a patch release. That's why I suggested catching (and rethrowing) the exception in the message handler, but also send the failed message to the failure channel (if configured).

**Reply by artembilan** [reply]
OK. That sounds like a plan, Gary.
Would you mind proposing a Pull Request?
Thanks

**Reply by garyrussell** [reply]
https://github.com/spring-projects/spring-integration/issues/8638

**Comment by yevsh** [comment]
@artembilan 
So to conclude this discussion, should i do what suggested above with RequestHandlerRetryAdvice?

I don't need to do explicit retry - just some other logic.

**Reply by artembilan** [reply]
Sure! You can use any other `MethodInterceptor` to catch those exceptions: https://docs.spring.io/spring-integration/docs/current/reference/html/messaging-endpoints.html#message-handler-advice-chain

**Comment by yevsh** [comment]
Still something not correct.
on handler:

```
       handler.setSync(true);
       handler.setAdviceChain(Collections.singletonList( myAdvice()));
```

advice:

```
 private ExpressionEvaluatingRequestHandlerAdvice myAdvice() {
        ExpressionEvaluatingRequestHandlerAdvice advice = new ExpressionEvaluatingRequestHandlerAdvice();

        advice.setFailureChannelName("failureChannel");
    
        advice.setTrapException(false); //true
        return advice;
    }
```

not triggered

**Reply by artembilan** [reply]
It is not clear if your `handler` is declared as a bean.
There is a logic like this:
```
	@Override
	protected final void onInit() {
		super.onInit();
		initAdvisedRequestHandlerIfAny();
		doInit();
	}
```
which is called from the `afterPropertiesSet()`.

**Comment by yevsh** [comment]
handler is a bean, what is not triggered is the `@ServiceActivator` of failureChannel when Exception is thrown.

I see now,

```
	private Object evaluateFailureExpression(Message<?> message, Exception exception) {

	     DestinationResolver<MessageChannel> channelResolver = getChannelResolver();
```
failed on
java.lang.IllegalArgumentException: 'beanFactory' must not be null

**Reply by artembilan** [reply]
Oh! Right. The `ExpressionEvaluatingRequestHandlerAdvice` instance must be configured as a bean.

**Reply by yevsh** [reply]
works good now. thanks.

Still feels a bit un intuitive adding advise just for catching and handling such  errors instead of using  directly

`handler.setSendFailureChannelName("failureChannel");`

but it's ok.

**Reply by artembilan** [reply]
That `handler.setSendFailureChannelName("failureChannel");` will work when we fix `KafkaTemplate` as I explained before with a `CompletableFuture.completeExceptionally(ex)` comment.
However it is indeed intuitive adding advise with Spring Integration concepts.

**Reply by yevsh** [reply]
Yes, I understand that. I also understand that fix will take time .So untill that will use the suggested approach.
